crl-release-22.1: nil pointer dereference

It appears to be possible to construct a database with 21.2 (36e8e921) that causes 22.1 (b66ae1e4) to panic with a nil pointer dereference on open. It appears as though L0Sublevels doesn't exist. If I recall correctly, that at least one point could happen if a compaction errored out. Did we forget to backport that? If so, it's also unclear why the compaction failed. Unfortunately, we don't have any of the EventListener logs here. I'll add that to the crossversion test.

```
    crossversion_test.go:72: Test directory: /tmp/TestMetaCrossVersion2746937505/001
    crossversion_test.go:73: Reproduction:
          go test -v -run 'TestMetaCrossVersion' --seed 1665774008718434026 --factor 10 --version 21.2,36e8e921,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test --version 22.1,b66ae1e4,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test --version 22.2,8f2c02f5,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test --version master,eec7375f,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.master.test

    crossversion_test.go:85: 0: 21.2,36e8e921,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test
    crossversion_test.go:85: 1: 22.1,b66ae1e4,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test
    crossversion_test.go:85: 2: 22.2,8f2c02f5,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test
    crossversion_test.go:85: 3: master,eec7375f,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.master.test
    crossversion_test.go:139: Running tests with version 36e8e921 with 1 initial state(s).
    crossversion_test.go:145:   Running test with version 36e8e921 with initial state <empty>.
    crossversion_test.go:139: Running tests with version b66ae1e4 with 10 initial state(s).
    crossversion_test.go:145:   Running test with version b66ae1e4 with initial state sha=36e8e921-seed=11375172571715449164-opts=standard-005(<empty>).
    crossversion_test.go:145:   Running test with version b66ae1e4 with initial state sha=36e8e921-seed=11375172571715449164-opts=random-020(<empty>).
    crossversion_test.go:145:   Running test with version b66ae1e4 with initial state sha=36e8e921-seed=11375172571715449164-opts=standard-008(<empty>).
    crossversion_test.go:145:   Running test with version b66ae1e4 with initial state sha=36e8e921-seed=11375172571715449164-opts=random-003(<empty>).
    crossversion_test.go:161: Metamorphic test failed: exit status 1
        Output:--- FAIL: TestMeta (1.29s)
            --- FAIL: TestMeta/standard-000 (0.02s)
                meta_test.go:352:
                    ===== SEED =====
                    10858758592825965365
                    ===== ERR =====
                    exit status 2
                    ===== OUT =====
                    --- FAIL: TestMeta (0.01s)
                    panic: runtime error: invalid memory address or nil pointer dereference [recovered]
                        panic: runtime error: invalid memory address or nil pointer dereference [recovered]
                        panic: runtime error: invalid memory address or nil pointer dereference
                    [signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x773511]

                    goroutine 19 [running]:
                    testing.tRunner.func1.2({0x96ab20, 0xf73d70})
                        /usr/local/go/src/testing/testing.go:1396 +0x24e
                    testing.tRunner.func1()
                        /usr/local/go/src/testing/testing.go:1399 +0x39f
                    panic({0x96ab20, 0xf73d70})
                        /usr/local/go/src/runtime/panic.go:884 +0x212
                    github.com/cockroachdb/pebble.Open.func1()
                        /home/jackson/go/src/github.com/cockroachdb/pebble/open.go:112 +0x175
                    panic({0x96ab20, 0xf73d70})
                        /usr/local/go/src/runtime/panic.go:890 +0x262
                    github.com/cockroachdb/pebble/internal/manifest.(*L0Sublevels).InUseKeyRanges(0x0, {0x0?, 0x0?, 0x0?}, {0x0, 0x0, 0x0})
                        /home/jackson/go/src/github.com/cockroachdb/pebble/internal/manifest/l0_sublevels.go:887 +0x51
                    github.com/cockroachdb/pebble.calculateInuseKeyRanges(0xc00017a140, 0xa25dc0, 0xc000181c80?, 0x6, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0})
                        /home/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:664 +0x95
                    github.com/cockroachdb/pebble.(*compaction).setupInuseKeyRanges(0xc00016a580)
                        /home/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:648 +0x85
                    github.com/cockroachdb/pebble.newFlush(0xc000004280, 0xc00017a140, 0x34?, {0x0, 0x0, 0x0}, 0xc00016e010)
                        /home/jackson/go/src/github.com/cockroachdb/pebble/compaction.go:639 +0x72a
```

```
                    ===== HISTORY =====
                    // INFO: [JOB 1] MANIFEST created 000001
                    // INFO: upgraded to format version: 002
                    // INFO: upgraded to format version: 003
                    // INFO: upgraded to format version: 004
                    // INFO: upgraded to format version: 005
```

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -33,6 +33,7 @@ jobs:
     - run: make test generate
 
   linux-crossversion:
+    if: ${{ false }} # disabled in PR 2018 until further investigation
     name: go-linux-crossversion
     runs-on: ubuntu-latest
     steps:

--- a/internal/metamorphic/options.go
+++ b/internal/metamorphic/options.go
@@ -347,13 +347,13 @@ func randomOptions(rng *rand.Rand) *testOptions {
 	return testOpts
 }
 
-func setupInitialState(dir string, testOpts *testOptions) error {
-	// Copy (vfs.Default,<initialStatePath>) to (testOpts.opts.FS,<dir>).
+func setupInitialState(dataDir string, testOpts *testOptions) error {
+	// Copy (vfs.Default,<initialStatePath>/data) to (testOpts.opts.FS,<dataDir>).
 	ok, err := vfs.Clone(
 		vfs.Default,
 		testOpts.opts.FS,
-		testOpts.initialStatePath,
-		dir,
+		vfs.Default.PathJoin(testOpts.initialStatePath, "data"),
+		dataDir,
 		vfs.CloneSync,
 		vfs.CloneSkip(func(filename string) bool {
 			// Skip the archive of historical files, any checkpoints created by
@@ -371,15 +371,15 @@ func setupInitialState(dir string, testOpts *testOptions) error {
 	// database (initialStatePath) could've had wal_dir set, or the current test
 	// options (testOpts) could have wal_dir set, or both.
 	fs := testOpts.opts.FS
-	walDir := fs.PathJoin(dir, "wal")
+	walDir := fs.PathJoin(dataDir, "wal")
 	if err := fs.MkdirAll(walDir, os.ModePerm); err != nil {
 		return err
 	}
 
-	// Copy <dir>/wal/*.log -> <dir>.
-	src, dst := walDir, dir
+	// Copy <dataDir>/wal/*.log -> <dataDir>.
+	src, dst := walDir, dataDir
 	if testOpts.opts.WALDir != "" {
-		// Copy <dir>/*.log -> <dir>/wal.
+		// Copy <dataDir>/*.log -> <dataDir>/wal.
 		src, dst = dst, src
 	}
 	return moveLogs(fs, src, dst)

--- a/internal/metamorphic/options_test.go
+++ b/internal/metamorphic/options_test.go
@@ -5,8 +5,6 @@
 package metamorphic
 
 import (
-	"os"
-	"path/filepath"
 	"testing"
 
 	"github.com/cockroachdb/pebble"
@@ -18,8 +16,9 @@ import (
 func TestSetupInitialState(t *testing.T) {
 	// Construct a small database in the test's TempDir.
 	initialStatePath := t.TempDir()
+	initialDataPath := vfs.Default.PathJoin(initialStatePath, "data")
 	{
-		d, err := pebble.Open(initialStatePath, &pebble.Options{})
+		d, err := pebble.Open(initialDataPath, &pebble.Options{})
 		require.NoError(t, err)
 		const maxKeyLen = 2
 		ks := testkeys.Alpha(maxKeyLen)
@@ -33,7 +32,6 @@ func TestSetupInitialState(t *testing.T) {
 		}
 		require.NoError(t, d.Close())
 	}
-	require.NoError(t, vfs.Default.MkdirAll(filepath.Join(initialStatePath, "wal"), os.ModePerm))
 	ls, err := vfs.Default.List(initialStatePath)
 	require.NoError(t, err)
 
@@ -44,7 +42,7 @@ func TestSetupInitialState(t *testing.T) {
 		initialStatePath: initialStatePath,
 		initialStateDesc: "test",
 	}
-	require.NoError(t, setupInitialState("", opts))
+	require.NoError(t, setupInitialState("data", opts))
 	copied, err := opts.opts.FS.List("")
 	require.NoError(t, err)
 	require.ElementsMatch(t, ls, copied)

